### PR TITLE
Use cryptographically generated passwords

### PR DESCRIPTION
**Pull request by sebastienros**
Seeing [an attempt to generate random passwords](https://github.com/dotnet/aspire/pull/2204) and other discussions about the recommended approach I am providing what I think is a better approach.

This could also be used to generate random ids but it's not part of this PR.
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2210)

### CODE REVIEW

File: src/Shared/PasswordGenerator.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Security.Cryptography;
+
+namespace Aspire;
+
+internal abstract class PasswordGenerator
```
Comment (eerhardt) [inline src/Shared/PasswordGenerator.cs:8]:
```suggestion
internal static class PasswordGenerator
```

File: tests/Aspire.Components.Common.Tests/Aspire.Components.Common.Tests.csproj [hunk 2]
```diff
@@ -5,6 +5,10 @@
     <IsTestSupportProject>true</IsTestSupportProject>
   </PropertyGroup>
 
+  <ItemGroup>
+    <Compile Include="..\..\src\Components\Common\PasswordGenerator.cs" Link="PasswordGenerator.cs" />
+  </ItemGroup>
+
```
Comment (eerhardt) [inline tests/Aspire.Components.Common.Tests/Aspire.Components.Common.Tests.csproj:11]:
```suggestion
```

1. This is unnecessary.
2. This link is pointing to a place that doesn't exist. (Common vs Shared).

File: src/Shared/PasswordGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (eerhardt) [inline src/Shared/PasswordGenerator.cs:1]: Why is this in the `Shared` directory, but only used in 1 project? Can we just put it in the Aspire.Hosting folder?

File: src/Shared/PasswordGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+namespace Aspire;
+
+internal abstract class PasswordGenerator
+{
+    // Some chars are excluded:
+    // - prevent potential confusions, e.g., 0,o,O and i,I,l
+    // - exclude special chars which could interfere with command line arguments or connection strings, e.g. =,$,...
+
+    internal const string LowerCaseChars = "abcdefghjkmnpqrstvwxyz"; // exclude i,l,u,o
+    internal const string UpperCaseChars = "ABCDEFGHJKMNPQRSTVWXYZ"; // exclude I,L,U,O
```
Comment (eerhardt) [inline src/Shared/PasswordGenerator.cs:15]: Why are these characters being excluded?

File: src/Shared/PasswordGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        ArgumentOutOfRangeException.ThrowIfNegative(lowerCase);
+        ArgumentOutOfRangeException.ThrowIfNegative(upperCase);
+        ArgumentOutOfRangeException.ThrowIfNegative(digit);
+        ArgumentOutOfRangeException.ThrowIfNegative(special);
+
+        var length = lowerCase + upperCase + digit + special;
+
+        ArgumentOutOfRangeException.ThrowIfZero(length);
+
+        Span<char> chars = stackalloc char[length];
```
Comment (eerhardt) [inline src/Shared/PasswordGenerator.cs:39]: We might want to add an assert that the length doesn't get crazy big, say over 128?

File: src/Shared/PasswordGenerator.cs (hunk 4 above)
Comment (sebastienros) [reply src/Shared/PasswordGenerator.cs:15]: It can lead to unfriendly words.

File: src/Aspire.Hosting/Utils/PasswordGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+internal static class PasswordGenerator
+{
+    // Some chars are excluded:
+    // - prevent potential confusions, e.g., 0,o,O and i,I,l
+    // - exclude special chars which could interfere with command line arguments or connection strings, e.g. =,$,...
+
+    internal const string LowerCaseChars = "abcdefghjkmnpqrstuvwxyz"; // exclude i,l,o
+    internal const string UpperCaseChars = "ABCDEFGHJKMNPQRSTUVWXYZ"; // exclude I,L,O
+    internal const string DigitChars = "0123456789";
+    internal const string SpecialChars = "-_#@!./:[]{}+,*()~"; // exclude &<>=;,`'^%$
```
Comment (eerhardt) [inline src/Aspire.Hosting/Utils/PasswordGenerator.cs:18]: The comments say to exclude a comma `,` but it is included in the string.

File: src/Aspire.Hosting/Utils/PasswordGenerator.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        ArgumentOutOfRangeException.ThrowIfZero(length);
+
+        Debug.Assert(length <= 128, "password too long");
+
+        Span<char> chars = stackalloc char[length];
+
+        RandomNumberGenerator.GetItems<char>(LowerCaseChars, lowerCase).CopyTo(chars);
+        RandomNumberGenerator.GetItems<char>(UpperCaseChars, upperCase).CopyTo(chars[lowerCase..]);
+        RandomNumberGenerator.GetItems<char>(DigitChars, digit).CopyTo(chars[(lowerCase + upperCase)..]);
+        RandomNumberGenerator.GetItems<char>(SpecialChars, special).CopyTo(chars[(lowerCase + upperCase + digit)..]);
```
Comment (eerhardt) [inline src/Aspire.Hosting/Utils/PasswordGenerator.cs:47]:
```suggestion
        RandomNumberGenerator.GetItems(LowerCaseChars, chars.Slice(0, lowerCase));
        RandomNumberGenerator.GetItems(UpperCaseChars, chars.Slice(lowerCase, upperCase));
        RandomNumberGenerator.GetItems(DigitChars, chars.Slice(lowerCase + upperCase, digit)));
        RandomNumberGenerator.GetItems(SpecialChars, chars.Slice(lowerCase + upperCase + digit, special));
```

No need in creating intermediate arrays.

File: src/Aspire.Hosting/Utils/PasswordGenerator.cs (hunk 6 above)
Comment (sebastienros) [reply src/Aspire.Hosting/Utils/PasswordGenerator.cs:18]:
```suggestion
    internal const string SpecialChars = "-_#@!./:[]{}+*()~"; // exclude &<>=;,`'^%$
```

**Review by eerhardt — APPROVED** [review]
Looks fine to me. Would be good to get @mitchdenny to review.

File: src/Aspire.Hosting/Utils/PasswordGenerator.cs [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+        ArgumentOutOfRangeException.ThrowIfZero(length);
+
+        Debug.Assert(length <= 128, "password too long");
+
+        Span<char> chars = stackalloc char[length];
+
+        RandomNumberGenerator.GetItems(LowerCaseChars, chars.Slice(0, lowerCase));
+        RandomNumberGenerator.GetItems(UpperCaseChars, chars.Slice(lowerCase, upperCase));
+        RandomNumberGenerator.GetItems(DigitChars, chars.Slice(lowerCase + upperCase, digit)));
```
Comment (eerhardt) [inline src/Aspire.Hosting/Utils/PasswordGenerator.cs:46]:
```suggestion
        RandomNumberGenerator.GetItems(DigitChars, chars.Slice(lowerCase + upperCase, digit));
```

File: tests/Aspire.Hosting.Tests/Utils/PasswordGeneratorTests.cs [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+        Assert.Throws<ArgumentOutOfRangeException>(() => PasswordGenerator.GeneratePassword(0, 0, 0, 0));
+    }
+
+    [Theory]
+    [InlineData(1)]
+    [InlineData(3)]
+    [InlineData(10)]
+    public void IncludesLowerCase(int length)
+    {
+        var password = PasswordGenerator.GeneratePassword(length, 0, 0, 0);
```
Comment (mitchdenny) [inline tests/Aspire.Hosting.Tests/Utils/PasswordGeneratorTests.cs:27]: This is not length, this is the # of lower case chars.

Comment (mitchdenny) [reply tests/Aspire.Hosting.Tests/Utils/PasswordGeneratorTests.cs:27]: Oh I see what you are demonstrating. :)

**Comment by mitchdenny** [comment]
It is definitely better than what we have today :)

One thing that I thought of and I don't know if it is really a practical issue for us given the ephemeral nature of the passwords we are generating, is that this code would always seem to be in the form:

[lower case chars set][upper case chars set][digit chars set][special chars set]

This reduces the possible set of characters for each individual character position which reduces the possible combination of passwords. To give a concrete example, here is the dictionary of characters we have by category:

|Set|Characters|Quantity|
|-|-|-|
| Lowercase | `abcdefghjkmnpqrstuvwxyz` | 23 |
| Uppercase | `ABCDEFGHJKMNPQRSTUVWXYZ` | 23 |
| Digits | `0123456789` | 10 |
| Special chars | `-_#@!./:[]{}+*()~` | 17 |

So, if we call the following code (like we do in SQL Server):

```csharp
GeneratePassword(6, 6, 2, 2)
```

The number of combinations is:

23 * 23 * 23 * 23 * 23 * 23 * 23 * 23 * 23 * 23 * 23 * 23 * 10 * 10 * 17 * 17 = 6.33332646085387E+20

Compare this with a GUID which is 1.208925819614629174706176e+24

If we instead allowed the characters to be in any position and the options on the password meant the minimum number of characters of a given set and didn't correlate to password length or where they were found.

**Comment by eerhardt** [comment]
> If we instead allowed the characters to be in any position

They get shuffled at the end. So the characters can be in any position.

**Comment by mitchdenny** [comment]
Ah i missed that ignore everything I said.

**Review by mitchdenny — APPROVED** [review]

**Pull request merged**